File: leetcode/algorithms/array/3sum.ts
/*
	leetcode
	category: "Algorithms"
	problem title: "3Sum"
*/

function threeSum(nums: number[]): number[][] {
    nums.sort((a,b) => a-b);
    const triplets: number[][] = [];
    
    let i: number = 0;
    
    while(i<nums.length-2){
        const target = 0-nums[i];
        
        let l: number = i+1,h: number = nums.length-1;
        
        while(h-l>0){
            
            const intSum: number = nums[h]+nums[l];
            
            if(intSum>target){
                h--;
            } else if(intSum<target){
                l++;
            } else {
                triplets.push([nums[h],nums[l],nums[i]])
                while(l<nums.length-1 && nums[l]===nums[l+1]){
                    l++
                }
                while(h>0 && nums[h]===nums[h-1]){
                    h--;
                }
                l++;
                h--;
            }
            
        }
        
        while(i<nums.length-2 && nums[i]===nums[i+1]){
            ++i;
        }
        ++i   
    }
    
    return triplets
    

};

File: leetcode/algorithms/array/3sum.py
class Solution(object):
    def threeSum(self, nums):
        target = 0;
        nums.sort();
        
        triplets = []
        
        i=0;
        
        while i<len(nums)-2:
            v = target - nums[i];
            x=i+1
            y=len(nums)-1;
            while(y-x>=1):
                if(nums[x]+nums[y]>v):
                    y=y-1;
                elif(nums[x]+nums[y]<v):
                    x=x+1;
                else:
                    triplets.append([nums[i],nums[x],nums[y]]);
                    while(x<len(nums)-2 and nums[x]==nums[x+1]):
                        x=x+1
                    x=x+1;
                    while(y>i and nums[y]==nums[y-1]):
                        y=y-1;
                    y=y-1;
            
            while(i<len(nums)-2 and nums[i]==nums[i+1]):
                i=i+1;
            i=i+1;
        
        return triplets;